### PR TITLE
add mlp, lstm, transformer

### PR DESCRIPTION
**Pull request by nikita-0209**
As discussed, 
1. Removed `base2proxy()` and added a `preprocess()` function to the three models. 
2. I haven't added the CNN implementation as its accuracy was not that great. However, after further experiments if we feel that CNN is a better choice, I'll add that (should not be a major change)
3. For now, I have hard-coded the model parameters (based on the best performing model in the experiments I conducted). We can modify this once we configure hydra
4. The different model take inputs in different formats:
`x: DNA Sequence` `xlens: length of each sequence in batch` 
(a) `MLP`: `model(x)`
(b) `LSTM`: `model(x, xlens)`
(c) `Transformer`: `model(x, mask)`
To take care of this, for now, I have added an `if` clause in the acquisition file. Certainly there are better ways to handle this (which I cannot think of at the moment). In case you have any ideas, let me know.

### CODE REVIEW

File: acquisition.py [hunk 1]
```diff
@@ -2,113 +2,128 @@
...
+            if self.config.proxy.model.lower() == "mlp":
+                outputs = self.proxy.model(inputs)
+            elif self.config.proxy.model.lower() == "lstm":
+                outputs = self.proxy.model(inputs, input_afs_lens)
+            elif self.config.proxy.model.lower() == "transformer":
+                outputs = self.proxy.model(inputs, None)
+
         return outputs
 
     def base2af(self, state):
```
Comment (manhbao-nguyen) [inline acquisition.py]: I agree that base2af could be removed because the input format for the acquisition function is going to be the input format of the proxy. So instead of implementing a base2af for each acquisition function, we can just take "preprocess" from each proxy !

File: proxy.py [hunk 2]
```diff
@@ -229,36 +239,60 @@ def check_convergence(self):
...
     @abstractmethod
     def evaluate(self, data):
         self.model.eval()
         with torch.no_grad():
             output = self.model(data).cpu().detach().numpy()
-            return output  
+            return output
 
     @abstractmethod
     def base2proxy(self, state):
```
Comment (Baobabtou) [inline proxy.py:251]: This is preprocess now, right ?

File: acquisition.py (hunk 1 above)
Comment (nikita-0209) [reply acquisition.py]: Yeah I agree. So I should delete `base2af()`? Ah, I don't think I understand the second part of your comment. Do you suggest explicitly calling the `preprocess()` function? In the current implementation, I am calling `forward()` and `preprocess()` is called implicitly inside `forward()`

Comment (manhbao-nguyen) [inline acquisition.py]: I suggest you delete  "base2af"  as it is not used. If there needs to be a specific function to preprocess for a certain acquisition function, we'll create it in the future and it will use proxy.model.preprocess, but no need for this simple AF :)

File: proxy.py (hunk 2 above)
Comment (manhbao-nguyen) [inline proxy.py]: Let's remove base2proxy as well (the preprocess method that replaces it is in the model(nn.Module) class directly and not the child Proxy class)

File: proxy.py [hunk 3]
```diff
@@ -270,49 +304,105 @@ def test(self, te):
...
+        return F.mse_loss(output[:, 0], targets.float())
+
     def check_convergence(self):
         super().check_convergence()
 
     def evaluate(self, data):
         return super().evaluate(data)
-  
+
     def base2proxy(self, state):
```
Comment (manhbao-nguyen) [inline proxy.py]: remove base2proxy from all proxy child classes

File: proxy.py [hunk 4]
```diff
@@ -402,62 +496,192 @@ def __init__(self, activation_func):
...
+        self.out_dim = 1
+        self.transformerCall = transformerCall
 
-        # initialize module lists
-        self.lin_layers = nn.ModuleList(self.lin_layers)
-        self.activations = nn.ModuleList(self.activations)
-        #self.norms = nn.ModuleList(self.norms)
-        self.dropouts = nn.ModuleList(self.dropouts)
+        if self.transformerCall == False:
+            self.hidden_layers = [1024, 1024, 1024, 1024, 1024]
```
Comment (alexhernandezgarcia) [inline proxy.py:521]: It would be nice if the layer dimensionality could be selected via configuration, but not critical to add before merging.

File: acquisition.py [hunk 5]
```diff
@@ -2,113 +2,97 @@
...
-        
-        return input_proxy.to(self.device)[0]
-    
-
+            if self.config.proxy.model.lower() == "mlp":
+                outputs = self.proxy.model(inputs)
+            elif self.config.proxy.model.lower() == "lstm":
+                outputs = self.proxy.model(inputs, input_afs_lens)
+            elif self.config.proxy.model.lower() == "transformer":
+                outputs = self.proxy.model(inputs, None)
```
Comment (alexhernandezgarcia) [inline acquisition.py:96]: It's a good practice to always include an `else` and raise an error (e.g. not defined) if necessary to catch potential issues.

**Comment by alexhernandezgarcia** [comment]
> The different model take inputs in different formats:
x: DNA Sequence xlens: length of each sequence in batch
(a) MLP: model(x)
(b) LSTM: model(x, xlens)
(c) Transformer: model(x, mask)

Ideally all models should share the same interface, as you said. On possible way is add all possible arguments (`x, xlens, mask`) to all cases and just pass None when unused.

It's not critical, but I would add [an issue](https://github.com/alexhernandezgarcia/ActiveLearningPipeline/issues) before merging to keep track of this.

Otherwise, you may go ahead and merge.
